### ISSUE TITLE
Support for more 3D primitives

### ISSUE THREAD

**Issue by jleibs** [rerun-io/rerun#1361]
It would be nice to be able to log the common 2d/3d geometry building blocks in order to enable building up simple 3d models without going all the way to fully specified meshes.
 - Rect / Box
 - Circle / Sphere
 - Triangle / Cone
 - Cylinder

Maybe more, but [good enough for VRML, good enough for me?](https://web.cs.wpi.edu/~matt/courses/cs563/talks/vrml/primitive.html)

These should all consistently support components for Orientation, Color, and Fill.

**Comment by emilk** [comment]
I think we should add a minimal version of this ASAP - it can be something that just generates meshes during scene construction for now. This is very limiting at the moment, and we need this to test other things like transparency, 2D layering, and styling.

**Comment by emilk** [comment]
An interesting note: a `Rectangle` component is useful both in 2D and 3D.

For 2D we support it via specifying its min/max corners, or center and size, and some other options. We will also add support for 2D transforms (https://github.com/rerun-io/rerun/issues/349).

For 3D we may want the same 2D interface plus a 3D transform, treating the 2D rect as a 3D shape with an implicit Z=0.

**Comment by emilk** [comment]
Rect/Box we have, but needs styling (Separate issue).

Cylinders and capsules I think are very useful (e.g. limbs)

2D circles/discs are very useful. So are spheres. What is the difference with point?

Point radii can be auto and ui-based, but spheres/circles _need_ a real radius.
A sphere/circle would squash with non-uniform transforms.
Points are using a faster, less accurate, renderer.

**Comment by jeongseok-meta** [comment]
It would be great to have a way to visualize the grid or ground. This could be achieved through a loggable object or simply as a visual artifact.

**Comment by mont1004** [comment]
Hi there. 

Just curious, has there been any progress or updates on this topic? Would love to know what's going on.
I personally hope for support of an ellipsoid representation in physical units, which can be beneficial for expressing the ambiguity in, for example, camera pose estimation.

**Comment by abey79** [comment]
@mont1004 Unfortunately the next really will have little new on this front, but this is still something that's high on our priority list. Do you have any particular primitive in mind that you would need most?

**Comment by mont1004** [comment]
@abey79 Thank you for the quick reply!

My team has been developing Visual SLAM, and we've been eyeing the rerun as a top candidate for our new visualizer. As I mentioned earlier, we currently represent the covariance of the estimated camera poses with ellipsoids. Having ellipsoids that can be scaled and compared in size with one another would be very helpful, as this capability is crucial for us.

**Comment by pablovela5620** [comment]
Want to add like #6012 that adding text to an image, similar to OpenCV would be appreciated, oftentimes I don't want to log something in textlog/textdocument as I would like it to show in the image itself. As a work around I log a point2d or linestrip with 0 radius and add a label. This works, but I can't easily change the size of the font like cv2.drawtext would allow

**Comment by nikolausWest** [comment]
@pablovela5620: Are you adding the text to a particular place in the image (i.e. tying it to a geometric location), or is the text more of a description of the image (i.e. it could alternatively be below the image for instance)?

**Comment by pablovela5620** [comment]
Generally to a geometric location, for example detecting someone's face and placing the text next to or above their detected face. Then telling them to either get closer or farther from the screen or to move their face to a specific location. Below an image could also be helpful (though I think this would already be possible with a textdoc + blueprint?)

**Comment by nikolausWest** [comment]
Rays and panes as per #2179 would be great

**Comment by emilk** [comment]
For the 3D primitives, we should have a style option to either render them as solid, as wireframe, or both. The spatial space-view can convert them to a re_renderer mesh and/or lines as needed.

To keep things simple, I suggest we use the `Transform3D ` archetype to orient and position the shapes. This means that e.g. a cylinder would always have its base at the origin, and be defined by `Cylinder { radius, height }`.

Which axis should "height" be along? I doesn't really matter, but once we picked something we should stick with it forever, and also use it for all other shapes (i.e. the axis along which a capsule and cylinder stretches). It's worth spending some time investigating what is the most common choice in other visualization tools.

We should still have nice helpers to construct these primitives. For instance, defining a capsule by two points and a radius (a fat line segment) but then logging that as a single `component Capsule { radius, length }` plus an `Transform3D`. This would mean an `archetype Capsule` would also _contain_ `archetype.Transform3D` in a sense, which may require some more codegen design/thinking.

For the 3D ellipsoids: I did some preliminary experimentation with these in https://github.com/rerun-io/rerun/pull/4991

**Comment by teh-cmc** [comment]
> We should still have nice helpers to construct these primitives. For instance, defining a capsule by two points and a radius (a fat line segment) but then logging that as a single `component Capsule { radius, length }` plus an `Transform3D`. This would mean an `archetype Capsule` would also _contain_ `archetype.Transform3D` in a sense, which may require some more codegen design/thinking.

We use the `OutOfTreeTransform3D` component for this today; it has worked well for `Asset3D`.

**Comment by emilk** [comment]
A less important, but very nice, primitive, is the tapered capsule:

![image](https://github.com/rerun-io/rerun/assets/1148717/ed477fe3-7693-40a5-a88a-b16e442b0ff2)

**Comment by kpreid** [comment]
Progress update:

* #6853 has added the `Ellipsoids` archetype.
* #6953 has added general mechanisms for solid/filled surface rendering (so our 3D geometric shapes can be solid) and applied that to `Boxes3D` and `Ellipsoids`.

More 3D shapes to come!

**Comment by emilk** [comment]
# What is the main axis of a cylinder/capule?
URDF: ???
## X
* PhysX: https://docs.nvidia.com/gameworks/content/gameworkslibrary/physx/guide/Manual/Geometry.html#capsules

## Y Main axis
* VRML: https://www.wiley.com/legacy/compbooks/vrml2sbk/ch03/03fig05.htm
* three.js: https://github.com/mrdoob/three.js/blob/master/src/geometries/CylinderGeometry.js

**Comment by emilk** [comment]
A lot of 3D libraries have a configurable main-axis, that the user can set to any Vec3. I think we should provide the same. That way a user can easily configure where a cylinder points, without having to calculate a quaternion and read our docs about what our default is.

So I suggest we introduce a `components.MainAxis3: datatypes.Vec3` that defaults to `+Z`.
A user can still add any transform on top of it (including rotation).

**Comment by 0x7CFE** [comment]
Hello,

I'd like to contribute by implementing `Circles2D` primitive since I need it to visualize my ML model.

Currently I'm using an app I wrote in [iced](https://github.com/iced-rs/iced) where it all looks something like this (the app in action can be seen on [this Youtube video](https://youtu.be/n27tv8KyUDY)):
<img width="50%" alt="Image" src="https://github.com/user-attachments/assets/4dce85b3-e5cd-4a12-8a0c-0bde10f6ffb9" />

I was able to use `Points2D` for the top-left view, but the remaining ones are tricky. Since `Circles2D` is probably not a top priority, I am wondering, how hard it is to implement? If it's not that hard maybe I would help this awesome project by implementing the primitive myself.

**Comment by abey79** [comment]
@0x7CFE thanks for asking.

To be honest, it is a surprisingly involved task, no so much because of the actual rendering (this could be done e.g. by piggy-backing on the existing `LineStrip2D` rendering, though there might be better ways), but because the features is about introducing a new piece in Rerun's data model. That includes discussing/designing the archetype, setting up codegen for the various SDKs, producing example snippets, etc. It also involves ensuring that all the relevant views (in this case, the spatial 2d and map view) add support for the new archetype. You can have a taste of that by looking at some of the PRs referenced above.

That being said, it is certainly not an impossible task, and is mainly about retracing existing pathways in the code base. I don't mean to discourage a contribution here, but just set the expectations straight :)

**Comment by 0x7CFE** [comment]
@abey79, thank you for the answer.

After a few experiments I ended up with the following: 

<img width="50%"  alt="Image" src="https://github.com/user-attachments/assets/8e13a194-5405-45bd-9956-5e894b3003bb" />

Here I'm using `Boxes2D` with half-sizes of 0.01 (doesn't allow me to keep them 0) and radii that essentially makes them circles. I placed boxes below `Points2D` to simulate transparency (it doesn't work for some reason, even if I set it like `.with_colors([Color::from_unmultiplied_rgba(255, 255, 255, 128)])`. I kinda like the effect when I select `detectors` in the left pane and it causes all "boxes" to be outlined. However, since boxes are placed below the points, it's hard to keep them interactive.

So the question is: instead of adding a whole new `Circles2D` primitive, would it be easier to just implement `FillMode` for `Boxes2D`? I am thinking about `FillMode::MajorWireframe` that should probably be interpreted as a box with empty/transparent fill. Even better if we'd be able to set different colors for the edges and the fill, so I could replicate the outlined effect.

**Comment by abey79** [comment]
Very cool visual! 🤩 

Yeah alpha is currently not supported for points.

Reusing `FillMode` for 2d shapes seems slightly weird to me. However, we could add a `background_colors` optional component. (Having two components of the same type was impossible until recently, but this restriction has now been lifted).

I'm not sure the rendering side is ready for 2d shape transparency support, but an initial version could already support either 255 or 0 alpha to control the presence/absence of the background. It's also possible that some works need to be done on the rendering side for tessellation.

But @Wumpf should fact check all of the above, as I've been slightly out of touch with these parts recently.

**Comment by Wumpf** [comment]
Yeah, we haven't figure out our 2d draw order story yet which is why we don't have transparency on 2D things yet. It got recently unblocked since the renderer finally performs distance based sorting, but 2D draw order is still controlled via a fairly flaky z-buffer hack.
Doing a 255/0 hack isn't great imho since "not drawing" has very different semantics also for selection. I'd rather "just fix it finally" ;-)

And yes, fact checking that this screenshot looks pretty cool ✔️ 😁

**Comment by 0x7CFE** [comment]
Oh, I see. Honestly, `background_colors` would work perfectly fine for me. All things considered, what would be the proper/reasonable way to move forward?

P.S.: It looks even better once the clustering is solved 😁
<img width="50%" height="631" alt="Image" src="https://github.com/user-attachments/assets/4ae6adb2-2edd-4d6f-bf80-b4ff9d60ea22" />

**Comment by abey79** [comment]
@0x7CFE We've had a design discussion internally and we believe that adding an optional `background_colors` field to the `Boxes2d` is the right way to go. If the component is missing, no background is displayed. Also, initially it would only support alpha == 255.

If you are willing to make a contribution, then you can go forward with that.

From the top of my head, the overall process involves:
- add the field to `boxes2d.fbs`
- `pixi run codegen`
- update `visualizers/boxes2d.rs` to add support for rendering the background
- a pass on the various SDK (at least the python sdk) might be required
- update the `boxes3d_batch` set of snippet to include a background colour (for documentation purposes) 

Happy to assist with the details.
